### PR TITLE
[Numpy] Numpy copysign

### PR DESCRIPTION
**Pull request by hzfan**
## Description ##
Numpy compatible [copysign](https://docs.scipy.org/doc/numpy/reference/generated/numpy.copysign.html)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- add copysign, including doc and test

## Comments ##
- Thank @reminisce @haojin2  for guidance and review.

### CODE REVIEW

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 1]
```diff
@@ -150,6 +150,26 @@ Example::
...
+.set_attr<nnvm::FInplaceOption>("FInplaceOption",
+  [](const NodeAttrs& attrs){
+    return std::vector<std::pair<int, int> >{{0, 1}};
+  })
+.set_attr<FResourceRequest>("FResourceRequest",
+  [](const NodeAttrs& attrs) {
+    return std::vector<ResourceRequest>{ResourceRequest::kTempSpace};
+  })
+.set_attr<FCompute>("FCompute<cpu>", BinaryBroadcastBackwardUseIn<cpu, mshadow_op::copysign_grad,
+                                                              mshadow_op::copysign_rgrad>);
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]: Alignment:
```c++
.set_attr<FCompute>("FCompute<cpu>", BinaryBroadcastBackwardUseIn<cpu, mshadow_op::copysign_grad,
                                                                  mshadow_op::copysign_rgrad>);
```

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 2]
```diff
@@ -182,5 +202,25 @@ MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_rpower_scalar)
...
+.add_argument("scalar", "float", "scalar value")
+.set_attr_parser([](NodeAttrs *attrs) { attrs->parsed = std::stod(attrs->dict["scalar"]); })
+.set_attr<FCompute>("FCompute<cpu>",
+                    BinaryScalarOp::Backward<cpu, mshadow_op::copysign_grad>);
+
+MXNET_OPERATOR_REGISTER_BINARY(_backward_npi_rcopysign_scalar)
+.add_argument("scalar", "float", "scalar value")
+.set_attr_parser([](NodeAttrs *attrs) { attrs->parsed = std::stod(attrs->dict["scalar"]); })
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Backward<
+  cpu, mshadow_op::rcopysign_grad>);
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]: nit:
I would prefer
```c++
.set_attr<FCompute>("FCompute<cpu>",
                    BinaryScalarOp::Backward<cpu, mshadow_op::rcopysign_grad>);
```

File: src/operator/numpy/np_elemwise_broadcast_op.cu [hunk 3]
```diff
@@ -42,6 +42,12 @@ NNVM_REGISTER_OP(_npi_mod)
 NNVM_REGISTER_OP(_npi_power)
 .set_attr<FCompute>("FCompute<gpu>", BinaryBroadcastCompute<gpu, mshadow_op::power>);
 
+NNVM_REGISTER_OP(_npi_copysign)
+.set_attr<FCompute>("FCompute<gpu>", BinaryBroadcastCompute<gpu, mshadow_op::copysign>);
+
+NNVM_REGISTER_OP(_backward_npi_copysign)
+.set_attr<FCompute>("FCompute<gpu>", BinaryBroadcastBackwardUseIn<gpu, mshadow_op::copysign_grad,
+                                                              mshadow_op::copysign_rgrad>);
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cu]: Alignment here too

File: src/operator/numpy/np_elemwise_broadcast_op.cc (hunk 2 above)
Comment (haojin2) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: Same for all other similar places.

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 4]
```diff
@@ -182,5 +202,25 @@ MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_rpower_scalar)
...
+MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_copysign_scalar)
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Compute<cpu, mshadow_op::copysign>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_npi_copysign_scalar"});
+
+MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_rcopysign_scalar)
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Compute<cpu, mshadow_op::rcopysign>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_npi_rcopysign_scalar"});
+
+MXNET_OPERATOR_REGISTER_BINARY(_backward_npi_copysign_scalar)
+.add_argument("scalar", "float", "scalar value")
```
Comment (sxjscience) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]: It seems we can use `MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR`

File: tests/python/unittest/test_numpy_op.py [hunk 5]
```diff
@@ -149,6 +149,66 @@ def hybrid_forward(self, F, a):
...
+            for hybridize in [True, False]:
+                for dtype in types:
+                    test_copysign = TestCopysign()
+                    if hybridize:
+                        test_copysign.hybridize()
+                    rtol = 1e-3
+                    atol = 1e-5
+                    a1_np = _np.array(_np.random.uniform(-1.0, 1.0, a1shape), dtype=dtype)
+                    a2_np = _np.array(_np.random.uniform(-1.0, 1.0, a2shape), dtype=dtype)
+                    a1 = mx.nd.array(a1_np).as_np_ndarray()
```
Comment (sxjscience) [inline tests/python/unittest/test_numpy_op.py]: We need to explicitly set the dtype: `a1 = mx.nd.array(a1_np, dtype=dtype).as_np_ndarray()

File: python/mxnet/ndarray/numpy/_op.py [hunk 6]
```diff
@@ -293,3 +293,46 @@ def power(x1, x2, out=None):
         This is a scalar if both x1 and x2 are scalars.
     """
     return _ufunc_helper(x1, x2, _npi.power, _np.power, _npi.power_scalar, _npi.rpower_scalar, out)
+
+
+@set_module('mxnet.ndarray.numpy')
+def copysign(x1, x2, out=None):
```
Comment (sxjscience) [inline python/mxnet/ndarray/numpy/_op.py]: I think we will not support the `where` argument. May need to keep the incompatible behaviors in a note.

File: src/operator/numpy/np_elemwise_broadcast_op.cc (hunk 2 above)
Comment (hzfan) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: fixed.

File: src/operator/numpy/np_elemwise_broadcast_op.cu (hunk 3 above)
Comment (hzfan) [reply src/operator/numpy/np_elemwise_broadcast_op.cu]: fixed.

File: tests/python/unittest/test_numpy_op.py (hunk 5 above)
Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: Yes. Have fixed it. Thank you!

File: src/operator/numpy/np_elemwise_broadcast_op.cc (hunk 4 above)
Comment (hzfan) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: Yes. Have fixed it. And Have added additional tests for scalar cases. Thank you!

File: python/mxnet/ndarray/numpy/_op.py (hunk 6 above)
Comment (hzfan) [reply python/mxnet/ndarray/numpy/_op.py]: Yes. Notes have been added in docs. See `_op.py`, `multiarray.py` and `_symbol.py`. Thank you!

File: tests/python/unittest/test_numpy_op.py [hunk 7]
```diff
@@ -1853,6 +1853,111 @@ def hybrid_forward(self, F, x):
...
+                    a1_np = _np.array(_np.random.uniform(-1.0, 1.0, a1shape), dtype=dtype)
+                    a2_np = _np.array(_np.random.uniform(-1.0, 1.0, a2shape), dtype=dtype)
+                    a1 = np.array(a1_np, dtype=dtype)
+                    a2 = np.array(a2_np, dtype=dtype)
+                    a1.attach_grad()
+                    a2.attach_grad()
+                    expected_np = _np.copysign(a1_np, a2_np)
+                    with mx.autograd.record():
+                        mx_out = test_copysign(a1, a2)
+                    assert mx_out.shape == expected_np.shape
```
Comment (sxjscience) [inline tests/python/unittest/test_numpy_op.py]: Test if mx_out.dtype matches expected_np.dtype. I think we can later add a utility to match the shape + dtype.

Comment (sxjscience) [reply tests/python/unittest/test_numpy_op.py]: Ignore my previous comments. I think we'd better check the return types after we support arbitrary dtype combinations in deepnumpy. So it's okay to not check the return types now.

Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: I agree that in the future we should do it, after more dtypes and casting is supported.

**Pull request merged**